optaplanner-examples/src/main/java/org/optaplanner/examples/common/experimental/impl/SequenceImpl.java

```diff
@@ -0,0 +1,117 @@
+/*
+ * Copyright 2021 Red Hat, Inc. and/or its affiliates.
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *      http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+package org.optaplanner.examples.common.experimental.impl;
+
+import java.util.NavigableSet;
+import java.util.Objects;
+import java.util.stream.Collectors;
+
+import org.optaplanner.examples.common.experimental.api.Sequence;
+
+public class SequenceImpl<ValueType_, DifferenceType_ extends Comparable<DifferenceType_>>
+        implements Sequence<ValueType_, DifferenceType_> {
+    private NavigableSet<ValueType_> consecutiveItemsSet;
+    private ValueType_ firstItem;
+    private ValueType_ lastItem;
+    private final ConsecutiveSetTree<ValueType_, ?, DifferenceType_> sourceTree;
+
+    protected SequenceImpl(ConsecutiveSetTree<ValueType_, ?, DifferenceType_> sourceTree, ValueType_ item) {
+        this.sourceTree = sourceTree;
+        this.consecutiveItemsSet = sourceTree.getItemSet().subSet(item, true, item, true);
+        this.firstItem = item;
+        this.lastItem = item;
+    }
+
+    protected SequenceImpl(ConsecutiveSetTree<ValueType_, ?, DifferenceType_> sourceTree,
+            NavigableSet<ValueType_> consecutiveItemsSet) {
+        this.sourceTree = sourceTree;
+        this.consecutiveItemsSet = consecutiveItemsSet;
+        this.firstItem = consecutiveItemsSet.first();
+        this.lastItem = consecutiveItemsSet.last();
+    }
+
+    @Override
+    public ValueType_ getFirstItem() {
+        return firstItem;
+    }
+
+    @Override
+    public ValueType_ getLastItem() {
+        return lastItem;
+    }
+
+    @Override
+    public NavigableSet<ValueType_> getItems() {
+        return consecutiveItemsSet;
+    }
+
+    @Override
+    public int getCount() {
+        return consecutiveItemsSet.size();
+    }
+
+    @Override
+    public DifferenceType_ getLength() {
+        return sourceTree.getSequenceLength(this);
+    }
+
+    protected boolean isEmpty() {
+        return consecutiveItemsSet.isEmpty();
+    }
```

Comment:
This is really bothering me. In the big scheme of things, this is irrelevant - but if you take a look at how much this takes out of all collector operations, this is about ~10 %. I really do not understand how this could be costing anything.

Anyway... we know the first item, we know the last item. Do we need to delegate to the collection if we know that we have a first and a last item?

Reply:
inlined and changed to getFirstItem() == getLastItem()